pytest_ethereum/linker.py

```diff
@@ -0,0 +1,70 @@
+import cytoolz
+from eth_utils import to_canonical_address, to_hex
+from eth_utils.toolz import assoc_in, pipe
+
+from ethpm import Package
+from pytest_ethereum.exceptions import LinkerError
+from pytest_ethereum.utils.linker import (
+    create_latest_block_uri,
+    create_new_deployment_data,
+    get_deployment_address,
+    insert_new_deployment,
+)
+
+
+def linker(strategy, package):
+    return pipe(package, *strategy)
+
+
+def deploy(contract_name, *args):
+    """
+    Returns a newly created package and contract address.
+    Will deploy the given contract_name, if data exists in package. If
+    a deployment is found on the current w3 instance, it will return that deployment
+    rather than creating a new instance.
+    """
+    return _deploy(contract_name, args)
+
+
+@cytoolz.curry
+def _deploy(contract_name, args, package):
+    deployments = package.deployments
+    if contract_name in deployments:
+        return package, package.deployments[contract_name].address
+
+    # Deploy new instance
+    factory = package.get_contract_factory(contract_name)
+    tx_hash = factory.constructor(*args).transact()
+    tx_receipt = package.w3.eth.waitForTransactionReceipt(tx_hash)
+    new_address = to_canonical_address(tx_receipt.contractAddress)
+    # Create manifest copy with new deployment instance
+    latest_block_uri = create_latest_block_uri(package.w3, tx_receipt)
+    deployment_data = create_new_deployment_data(contract_name, new_address, tx_receipt)
+    new_manifest = insert_new_deployment(package, deployment_data, latest_block_uri)
+    return Package(new_manifest, package.w3), new_address
```

Comment:
What if instead of returning the address you returned the contract instance?  Just realized that will be the thing that people will almost always want so maybe do it for them?